zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchTagStore.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2015-2018 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2.elasticsearch;
+
+import java.util.List;
+import java.util.Locale;
+import zipkin2.Call;
+import zipkin2.elasticsearch.internal.IndexNameFormatter;
+import zipkin2.elasticsearch.internal.client.Aggregation;
+import zipkin2.elasticsearch.internal.client.SearchCallFactory;
+import zipkin2.elasticsearch.internal.client.SearchRequest;
+import zipkin2.storage.TagStore;
+
+final class ElasticsearchTagStore implements TagStore {
+  static final String TAG = "sitetags";
+
+  final boolean searchEnabled;
+  final IndexNameFormatter indexNameFormatter;
+  final SearchCallFactory search;
+  final int namesLookback;
+  final List<String> tag;
+
+  ElasticsearchTagStore(ElasticsearchStorage es) {
+    this.search = new SearchCallFactory(es.http());
+    this.indexNameFormatter = es.indexNameFormatter();
+    this.searchEnabled = es.searchEnabled();
+    this.namesLookback = es.namesLookback();
+    this.tag = es.tag();
+  }
+
+  @Override public Call<List<String>> getKeys() {
```

Comment:
if you make an abstract base type for the TagStore, it will be less tempting to make a specific api for this. Then, your base type has a builder which assigns the list tagKeys and then returns that (wrapped in a call to fit the api, but used directly elsewhere.. for example, if asking for values of a tag not in the list, you just return empty). That would be my advice.. note: the builder could be StorageComponent.Builder as it would be unexpected for TagStore to be built independently